AEPMessaging/Sources/Messaging.swift

```diff
@@ -306,16 +394,141 @@ public class Messaging: NSObject, Extension {
+            guard expectedSurfaces.contains(proposition.propositionInfo.scope) else {
+                Log.debug(label: MessagingConstants.LOG_TAG,
+                          "Ignoring proposition where scope (\(proposition.propositionInfo.scope)) does not match one of the expected surfaces (\(expectedSurfaces)).")
+                continue
+            }
+
+            guard let rulesString = proposition.items.first?.data.content, !rulesString.isEmpty else {
+                Log.debug(label: MessagingConstants.LOG_TAG, "Skipping proposition with no in-app message content.")
+                continue
+            }
+
+            guard let parsedRules = rulesEngine.launchRulesEngine.parseRule(rulesString, runtime: runtime) else {
+                Log.debug(label: MessagingConstants.LOG_TAG, "Skipping proposition with malformed in-app message content.")
+                continue
+            }
+
+            var propInfo: [String: PropositionInfo] = [:]
+            if let messageId = parsedRules.first?.consequences.first?.id {
+                propInfo[messageId] = proposition.propositionInfo
+            }
+
+            isFeedConsequence = parsedRules.first?.consequences.first?.isFeedItem ?? false
+            if !isFeedConsequence {
+                // pre-fetch the assets for this message if there are any defined
+                rulesEngine.cacheRemoteAssetsFor(parsedRules)
+
+                // store reporting data for this payload
+                tempPropInfo.merge(propInfo) { _, new in new }
+            } else {
+                tempFeedsInfo.merge(propInfo) { _, new in new }
+            }
+
+            tempPropositions.append(proposition)
+            rules.append(contentsOf: parsedRules)
+        }
+
+        if !isFeedConsequence {
+            if clearExisting {
+                propositionInfo = tempPropInfo
+                inMemoryPropositions = tempPropositions
+            } else {
+                propositionInfo.merge(tempPropInfo) { _, new in new }
+                inMemoryPropositions.append(contentsOf: tempPropositions)
+            }
+
+            if persistChanges {
+                cachePropositions()
+            }
+        } else {
+            if clearExisting {
+                inMemoryFeeds.removeAll()
+                feedsInfo = tempFeedsInfo
+            } else {
+                feedsInfo.merge(tempFeedsInfo) { _, new in new }
+            }
+        }
+
+        return rules
+    }
+
+    private func isValidSurface(_ surfaceUri: String) -> Bool {
+        guard URL(string: surfaceUri) != nil else {
+            return false
+        }
+
+        return true
+    }
+
+    private func mergeFeedsInMemory(_ feeds: [String: Feed], requestedSurfaces: [String]) {
+        for surface in requestedSurfaces {
+            if feeds[surface] != nil {
+                inMemoryFeeds[surface] = feeds[surface]
+            } else {
+                inMemoryFeeds.removeValue(forKey: surface)
```

Comment:
do we need to synchronize access to these in-memory collections to prevent modifying them while iterating?  is access already synchronized at a higher level?